crypto/fipsmodule/evp/p_hkdf.c

```diff
@@ -179,6 +179,85 @@ static int pkey_hkdf_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2) {
   }
 }
 
+static int pkey_hkdf_ctrl_str(EVP_PKEY_CTX *ctx, const char *type,
+                              const char *value) {
+  if (strcmp(type, "mode") == 0) {
+    int mode;
+
+    if (strcmp(value, "EXTRACT_AND_EXPAND") == 0) {
+      mode = EVP_PKEY_HKDEF_MODE_EXTRACT_AND_EXPAND;
+    } else if (strcmp(value, "EXTRACT_ONLY") == 0) {
+      mode = EVP_PKEY_HKDEF_MODE_EXTRACT_ONLY;
+    } else if (strcmp(value, "EXPAND_ONLY") == 0) {
+      mode = EVP_PKEY_HKDEF_MODE_EXPAND_ONLY;
+    } else {
+      return 0;
+    }
+
+    return EVP_PKEY_CTX_hkdf_mode(ctx, mode);
+  }
+
+  if (strcmp(type, "md") == 0) {
+    OPENSSL_BEGIN_ALLOW_DEPRECATED
+    return EVP_PKEY_CTX_md(ctx, EVP_PKEY_OP_DERIVE, EVP_PKEY_CTRL_HKDF_MD,
+                           value);
+    OPENSSL_END_ALLOW_DEPRECATED
+  }
+
+  if (strcmp(type, "salt") == 0) {
+    // What if the salt contains a 0-byte?
```

Comment:
if it does contain an `\x00`, wouldn't `saltlen` be too short? is this something we can test?

Reply:
I'm not sure whether there is a "too short" condition on the salt here. Unfortunately, when we hit a `\x00` in the salt/string, there's no way to tell whether that's "really" where the consumer expected the salt/string to end.

Reply:
Right, you can't know that unless the expected length is passed in as a parameter.